agent/system_metrics.go

```diff
@@ -0,0 +1,256 @@
+// Regardless of if Start returns an error, Close should still be called
+// to close any opened connections, set the exit code, and handle any returned
+// errors. This can also be handled by cancelling the provided context, but
+// any errors will only be logged to the global error logs in this case.
+func (s *SystemMetricsCollector) Start(ctx context.Context) error {
+	var err error
+	s.id, err = s.client.CreateSystemMetricRecord(ctx, *s.taskOpts)
+	if err != nil {
+		return errors.Wrap(err, "problem creating cedar system metrics metadata object")
+	}
+
+	var streamingCtx context.Context
+	streamingCtx, s.streamingCancel = context.WithCancel(ctx)
+
+	for _, collector := range s.collectors {
+		stream, err := s.client.StreamSystemMetrics(streamingCtx, s.id, collector.Name(), metrics.StreamOpts{})
+		if err != nil {
+			s.streamingCancel()
+			s.streamingCancel = nil
+			return errors.Wrap(err, fmt.Sprintf("problem creating system metrics stream for id %s and metricType %s", s.id, collector.Name()))
+		}
+
+		ftdcCollector := ftdc.NewStreamingCollector(1, stream)
+		err = ftdcCollector.SetMetadata(collector.Metadata())
+		if err != nil {
+			s.streamingCancel()
+			s.streamingCancel = nil
+			return errors.Wrap(err, fmt.Sprintf("problem setting metadata on stream for id %s and metricType %s", s.id, collector.Name()))
+		}
+
+		s.wg.Add(1)
+		go s.timedCollect(streamingCtx, collector, stream, ftdcCollector)
+	}
+	go s.closeOnCancel(ctx, streamingCtx)
+
+	return nil
+}
+
+func (s *SystemMetricsCollector) timedCollect(ctx context.Context, mc MetricCollector, stream *metrics.SystemMetricsWriteCloser, collector ftdc.Collector) {
+	defer s.wg.Done()
+
+	timer := time.NewTimer(0)
+	defer timer.Stop()
+	defer s.catcher.Add(errors.Wrap(stream.Close(), fmt.Sprintf("problem closing system metrics stream for id %s and metricType %s", s.id, mc.Name())))
+	defer s.catcher.Add(errors.Wrap(ftdc.FlushCollector(collector, stream), fmt.Sprintf("problem flushing system metrics data collector for id %s and metricType %s", s.id, mc.Name())))
+	defer s.catcher.Add(errors.Wrap(recovery.HandlePanicWithError(recover(), nil, fmt.Sprintf("panic in system metrics stream for id %s and metricType %s", s.id, mc.Name())), ""))
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-timer.C:
+			data, err := mc.Collect()
+			if err != nil {
+				s.catcher.Add(errors.Wrapf(err, "problem collecting system metrics data for id %s and metricType %s", s.id, mc.Name()))
+				return
+			}
+			err = collector.Add(data)
+			if err != nil {
+				s.catcher.Add(errors.Wrapf(err, "problem adding system metrics data to collector for id %s and metricType %s", s.id, mc.Name()))
+				return
+			}
+			_ = timer.Reset(s.interval)
+		}
+	}
+}
+
+func (s *SystemMetricsCollector) closeOnCancel(outerCtx, streamingCtx context.Context) {
+	for {
+		select {
+		case <-outerCtx.Done():
+			grip.Error(s.cleanup())
+			return
+		case <-streamingCtx.Done():
```

Comment:
It seems like this should also perform cleanup. It's kind of odd for it to only cleanup if the global context cancels, but not if the streaming one cancels.